boot/assets.go

```diff
@@ -790,16 +794,40 @@ func observeSuccessfulBootAssetsForBootloader(m *Modeenv, root string, opts *boo
 		return nil, nil
 	}
 
-	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
-	for _, trustedAsset := range trustedAssets {
+	byAssetName := make(map[string][]string)
+	for trustedAsset := range trustedAssets {
 		assetName := filepath.Base(trustedAsset)
+		l, ok := byAssetName[assetName]
+		if !ok {
+			byAssetName[assetName] = []string{trustedAsset}
+		} else {
+			byAssetName[assetName] = append(l, trustedAsset)
+		}
+	}
 
-		// find the hash of the file on disk
-		assetHash, err := cache.fileHash(filepath.Join(root, trustedAsset))
-		if err != nil && !os.IsNotExist(err) {
-			return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
+
+	for assetName, trustedAssetsForName := range byAssetName {
+		var trustedAsset string
+		var assetHash string
+		foundTrustedAsset := false
+		for _, trustedAssetCandidate := range trustedAssetsForName {
+			assetHashFound, err := cache.fileHash(filepath.Join(root, trustedAssetCandidate))
+			if err != nil && !os.IsNotExist(err) {
+				return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+			}
+			if assetHashFound == "" {
+				continue
+			}
+			if foundTrustedAsset {
+				logger.Noticef("bootloader %s has several asset of the same name %s", whichBootloader, assetName)
+			}
+			foundTrustedAsset = true
+			trustedAsset = trustedAssetCandidate
+			assetHash = assetHashFound
```

Comment:
are there new tests for the case of multiple asset for a single name? per the comment above I'm not quite sure whether this works sensibly  or not? could we instead consider to have some control here or in the bootloaader so that we store the new clashing assets under a different name?

EFI/boot/grubx64.go  => assetName = grubx64.go
EFI/ubuntu/grubx64.go => assetName  = ubuntu:grubx64.go

or similar, perhaps?

Reply:
I will try that.